system/lib/wasmfs/backends/ignore_case_backend.cpp

```diff
@@ -0,0 +1,126 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+// This file defines the Ignore Case Backend of the new file system.
+// It is a virtual backend that normalizes all file paths to lower case.
+
+#include "backend.h"
+#include "file.h"
+#include "memory_backend.h"
+#include "wasmfs.h"
+
+namespace {
+
+std::string normalize(const std::string& name) {
+  std::string result = name;
+  for (auto& ch : result) {
+    ch = std::tolower(ch);
+  }
+  return result;
+}
+} // namespace
+
+namespace wasmfs {
+
+// Problem: Child entries are stored both in IgnoreCaseDirectory and in baseDirectory.
+//
+// Original name case preservation could be possible if MemoryDirectory::entries
+// were accesible. Then, we would store the original case there but search
+// ignoring case.
+//
+class IgnoreCaseDirectory : public MemoryDirectory {
+  using BaseClass = MemoryDirectory;
+  std::shared_ptr<Directory> baseDirectory;
+
+public:
+  IgnoreCaseDirectory(std::shared_ptr<Directory> base, backend_t proxyBackend)
+    : BaseClass(base->locked().getMode(), proxyBackend), baseDirectory(base) {}
+
+  std::shared_ptr<File> getChild(const std::string& name) override {
+    return BaseClass::getChild(normalize(name));
+  }
+
+  std::shared_ptr<DataFile> insertDataFile(const std::string& name,
+                                           mode_t mode) override {
+    auto name2 = normalize(name);
+    auto child = baseDirectory->insertDataFile(name2, mode);
+    insertChild(name2, child);
+    return child;
+  }
+
+  std::shared_ptr<Directory> insertDirectory(const std::string& name,
+                                             mode_t mode) override {
+    auto name2 = normalize(name);
+    auto baseChild = baseDirectory->insertDirectory(name2, mode);
+    auto child = std::make_shared<IgnoreCaseDirectory>(baseChild, getBackend());
+    insertChild(name2, std::static_pointer_cast<File>(child));
+    return child;
+  }
+
+  std::shared_ptr<Symlink> insertSymlink(const std::string& name,
+                                         const std::string& target) override {
+    auto name2 = normalize(name);
+    auto child = baseDirectory->insertSymlink(name2, target);
+    insertChild(name2, child);
+    return child;
+  }
+
+  int insertMove(const std::string& name, std::shared_ptr<File> file) override {
+    return BaseClass::insertMove(normalize(name), file);
+  }
+
+  bool removeChild(const std::string& name) override {
```

Comment:
Found a problem: `Directory::Handle::removeChild()` doesn't call `maintainsFileIdentity()` and `setParent()`: https://github.com/emscripten-core/emscripten/blob/main/system/lib/wasmfs/file.cpp#L195
For backends which maintain file identity, there would be no entry in `dcache` so no parent reset would happen.
Is this a bug?